### ISSUE TITLE
proposal: add iota of vector, unary sel

### ISSUE THREAD

**Issue by rsc** [robpike/ivy#192]
Two intertwined proposals.

**iota of vector**

Since `rho iota n` is n for scalar n, it makes sense for `rho iota x y z` to be `x y z`. That is, iota of a vector should produce a matrix with that vector as its shape. One possibility is to number the cells of the matrix in ravel order:

```
op newiota (x y) = x y rho iota x*y

newiota 3 4
 1  2  3  4
 5  6  7  8
 9 10 11 12
```

This is cute but I think less useful in practice than the second possibility, which is to make each cell be its coordinates in the matrix. That makes the idiom `(cond x) sel iota rho x` generalize from vectors to matrices:

```
op newiota (x y) = x y rho (iota x) o., (iota y)

newiota 3 4
(1 1) (1 2) (1 3) (1 4)
(2 1) (2 2) (2 3) (2 4)
(3 1) (3 2) (3 3) (3 4)
```

This second form is useful because it can be combined with sel to find coordinates of something in a matrix. For example where are the primes in `3 4 rho iota 12`?

```
x=3 4 rho iota 12

x
 1  2  3  4
 5  6  7  8
 9 10 11 12

x in 2 3 5 7 11
0 1 1 0
1 0 1 0
0 0 1 0

,x in 2 3 5 7 11
0 1 1 0 1 0 1 0 0 0 1 0

(,x in 2 3 5 7 11) sel ,newiota rho x
(1 2) (1 3) (2 1) (2 3) (3 3)
```

This is what Dyalog does. See also https://aplwiki.com/wiki/Index_Generator#Vector_arguments. 

**unary sel**

The idiom '(cond x) sel iota rho x' is pretty common for finding the indices where a condition is true.
Perhaps the unary form of sel should be exactly this, or more specifically:

    op sel x = (,x) sel ,iota rho x

This corresponds to the [APL Indices/Where function](https://aplwiki.com/wiki/Indices), and that definition is exactly Dyalog's.

With that you can select the indices of a specific condition in a matrix:

```
op sel x = (,x) sel ,newiota rho x

sel x==7
(2 3)
```

I am on the fence about whether sel is the right name. It's what I've used when defining this operator in my own programs, but I also like the suggestion of 'where' from the other APLs:

```
op where x = (,x) sel iota rho x

x = 3*iota 10

where x==15
5

x[where not x mod 5]
15 30
```

**Comment by rsc** [comment]
I remembered another reason for unary sel, and that's indexing into matrices. It is possible that this is one of the leading axis vs trailing axis inconsistencies (I don't fully understand all that yet), but given a boolean condition vector v, `v sel m` selects columns but you can use the unary sel in `m[sel v]` to select rows, instead of the longer `transp v sel transp m`. It's arguably clearer to write either `m[sel v]` or `m[; sel v]` than remember which the binary sel does.

```
op sel x = x sel iota rho x

m = 5 3 rho iota 15

m[; sel 1 1 0]
 1  2
 4  5
 7  8
10 11
13 14

m[sel 1 1 0 0 1;]
 1  2  3
 4  5  6
13 14 15

```

**Comment by bear8642** [comment]
Where is probably the better name - R calls it's similar function "[which](https://search.r-project.org/R/refmans/base/html/which.html)" invoking a simlar readibility ("which values?" / "there values where…" )

J uses your first newiota whilst APL uses the second newiota, however this then pushes you into mixed result Marshall mentions on the latest [ArrayCast episode ](https://www.arraycast.com/episodes/episode95-aoc) - it might be worth thinking about using K's [Odometer](https://wiki.k-language.dev/wiki/Odometer) approach to keep the result flat

**Comment by bear8642** [comment]
Your annoyance with the dyadic sel does appear to be due to leading vs trailing axis inconsistencies - leading axis was developed later than APL/360 and as such the ivy `sel` corresponds to *trailing* axis selection.

The leading axis `sel` in APL is `⌿` which (on a matrix) select the columns. On higher dimensional arrays, the leading axis functions apply from the outside - the *leading* part of the shape.

```APL
      m←5 3 ⍴ ⍳ 15

⍝ trailing axis 
      m[; ⍸ 1 1 0]
 1  2
 4  5
 7  8
10 11
13 14
      1 1 0/m 
 1  2
 4  5
 7  8
10 11
13 14

⍝ leading axis
      m[⍸ 1 1 0 0 1;]
 1  2  3
 4  5  6
13 14 15
      1 1 0 0 1⌿m 
 1  2  3
 4  5  6
13 14 15
```

**Comment by rsc** [comment]
Having the shape of the returned list of indices depend on the rank of the array being searched seems fine to me. The goal is to make the primitive useful, and returning separate r,c seems almost always more useful than an encoded ravel-order index. And of course if people want the ravel order index it is easy to search ,x instead of x.

I think the leading vs trailing issue is a separate concern from this specific proposal. Even if Ivy hypothetically switched which one sel uses, being able to choose the other explicitly in the form m[sel x;] or m[;sel x] is still helpful.

**Comment by bear8642** [comment]
Yes, separate r,c is probably more useful - thought you might want to know what other systems have come up with.

> being able to choose the other explicitly
Yes, selecting the axis to apply along is very useful - this why the [Rank operator](https://aplwiki.com/wiki/Rank_(operator)) is for, generalising the [bracket axis](https://aplwiki.com/wiki/Function_axis) notation.

**Comment by robpike** [comment]
I am a little lost here.

Ivy already has iota of a vector.

iota 1 2 3
(1 1 1) (1 1 2) (1 1 3)
(1 2 1) (1 2 2) (1 2 3)

rho iota 1 2 3
1 2 3

Are you saying that's wrong? It's consistent with what other APLs do.

      ⍴ ⍳ 1 2 3
1 2 3
      ⍳ 1 2 3
 1 1 1  1 1 2  1 1 3 
 1 2 1  1 2 2  1 2 3

**Comment by rsc** [comment]
> Ivy already has iota of a vector.

I somehow missed this fact. I am not sure how. Sorry about that. Then 'unary sel' (or 'unary where') is the only new part.
